File: Darts/src/burlton/dartzee/code/screen/Dartboard.java
package burlton.dartzee.code.screen;

import burlton.core.code.obj.SuperHashMap;
import burlton.core.code.util.Debug;
import burlton.dartzee.code.listener.DartboardListener;
import burlton.dartzee.code.object.ColourWrapper;
import burlton.dartzee.code.object.Dart;
import burlton.dartzee.code.object.DartboardSegment;
import burlton.dartzee.code.object.DartboardSegmentKt;
import burlton.dartzee.code.screen.game.GamePanelX01;
import burlton.dartzee.code.utils.*;

import javax.sound.sampled.*;
import javax.swing.*;
import java.awt.*;
import java.awt.event.MouseEvent;
import java.awt.event.MouseListener;
import java.awt.event.MouseMotionListener;
import java.awt.image.BufferedImage;
import java.net.URL;
import java.util.ArrayList;
import java.util.Random;

public class Dartboard extends JLayeredPane
					   implements MouseListener,
					   MouseMotionListener
{
	private static final ImageIcon DARTIMG = new ImageIcon(GamePanelX01.class.getResource("/dartImage.png"));
	
	private static SuperHashMap<String, URL> hmSoundNameToUrl = new SuperHashMap<>();
	
	private SuperHashMap<Point, DartboardSegmentKt> hmPointToSegment = new SuperHashMap<>();
	private SuperHashMap<String, DartboardSegmentKt> hmSegmentKeyToSegment = new SuperHashMap<>();
	
	private ArrayList<JLabel> dartLabels = new ArrayList<>();
	
	private DartboardListener listener = null;
	private Point centerPt = new Point(200, 200);
	private double diameter = 360;
	private boolean renderScoreLabels = false;
	
	private int dartCount = 0;
	
	private boolean simulation = false;
	
	//Cached things
	private DartboardSegmentKt lastHoveredSegment = null;
	private ColourWrapper colourWrapper = null;
	private Clip clip = null;
	
	public Dartboard()
	{
		setLayout(null);
		add(dartboardLabel, Integer.valueOf(-1));
	}
	public Dartboard(int width, int height) 
	{
		setSize(width, height);
		dartboardLabel.setSize(width, height);
		setLayout(null);
		add(dartboardLabel, Integer.valueOf(-1));
	}

	private BufferedImage dartboardImage = null;
	private final JLabel dartboardLabel = new JLabel();
	private JLabel dodgyLabel = new JLabel(); //You know what this is...
	
	public void addDartboardListener(DartboardListener listener)
	{
		this.listener = listener;
	}
	
	public void paintDartboard()
	{
		paintDartboard(null, true);
	}
	public void paintDartboard(ColourWrapper colourWrapper, boolean listen)
	{	
		int width = getWidth();
		int height = getHeight();
		
		Debug.append("Painting darboard. Dim[" + width + "," + height + "]");
		
		dartboardLabel.setSize(width, height);
		
		//Initialise/clear down variables
		this.colourWrapper = colourWrapper;
		centerPt = new Point(width/2, height/2);
		diameter = 0.7 * width;
		hmPointToSegment.clear();
		dartboardImage = new BufferedImage(width, height, BufferedImage.TYPE_INT_ARGB);
		
		//Construct the segments, populated with their points. Cache pt -> segment.
		for (int x=0; x<width; x++) 
		{
		    for (int y=0; y<height; y++) 
		    {
		    	Point pt = new Point(x, y);
		    	factoryAndCacheSegmentForPoint(pt);
		    }
		}
		
		Debug.append("Cached all points/segments.");
		
		//Render the actual image
		renderDartboardImage(width, height);
		
		addScoreLabels();
		
		//Now the dartboard is painted, add the mouse listeners
		if (listen)
		{
			ensureListening();
		}
	}
	
	private void renderDartboardImage(int width, int height)
	{
		int[] pixels = new int[width * height];
		int pixelIx = 0;
		Point pt = new Point();
		for (int y=0; y<height; y++)
	    {
	        for (int x=0; x<width; x++)
	        {
	        	pt.setLocation(x, y);

				DartboardSegmentKt segment = getSegmentForPoint(pt);
				Color colour = DartboardUtil.getColourForPointAndSegment(pt, segment, false, colourWrapper);
		    	
		    	if (colour != null)
		    	{
		    		pixels[pixelIx] = colour.getRGB();
		    	}
		    	
		    	pixelIx++;
	        }
	    }
		
		dartboardImage.setRGB(0, 0, width, height, pixels, 0, width);
		
		Debug.append("Created dartboardImage");
		
		dartboardLabel.setIcon(new ImageIcon(dartboardImage));
		dartboardLabel.repaint();
	}
	
	private void addScoreLabels()
	{
		if (!renderScoreLabels)
		{
			Debug.append("Not adding scores.");
			return;
		}
		
		//Get the height we want for our labels, which is half the thickness of the outer band
		double radius = diameter / 2;
		double outerRadius = DartboardUtil.UPPER_BOUND_OUTSIDE_BOARD_RATIO * radius;
		int lblHeight = (int)Math.round((outerRadius - radius)/2);
		
		Font fontToUse = getFontForDartboardLabels(lblHeight);
		
		for (int i=1; i<=20; i++)
		{
			//Create a label with standard properties
			JLabel lbl = new JLabel("" + i);
			lbl.setForeground(Color.WHITE);
			lbl.setBackground(DartsColour.TRANSPARENT);
			lbl.setOpaque(true);
			lbl.setHorizontalAlignment(SwingConstants.CENTER);
			lbl.setFont(fontToUse);
			
			//Work out the width for this label, based on the text
			FontMetrics metrics = factoryFontMetrics(fontToUse);
			int lblWidth = metrics.stringWidth("" + i) + 2;
			lbl.setSize(lblWidth, lblHeight);
			
			//Work out where to place the label
			ArrayList<Point> points = getPointsForSegment(i, DartboardSegment.TYPE_MISS);
			Point avgPoint = GeometryUtil.getAverage(points);
			int lblX = (int)avgPoint.getX() - (lblWidth / 2);
			int lblY = (int)avgPoint.getY() - (lblHeight / 2);
			lbl.setLocation(lblX, lblY);
			
			//Add to the screen
			add(lbl);
		}
	}
	private Font getFontForDartboardLabels(int lblHeight)
	{
		//Start with a fontSize of 1
		int fontSize = 1;
		Font font = new Font("Trebuchet MS", Font.PLAIN, fontSize);
		
		//We're going to increment our test font 1 at a time, and keep checking its height
		Font testFont = font;
		FontMetrics metrics = factoryFontMetrics(testFont);
		int fontHeight = metrics.getHeight();
		
		while (fontHeight < lblHeight - 2)
		{
			//The last iteration succeeded, so set our return value to be the font we tested.
			font = testFont;
			
			//Create a new testFont, with incremented font size
			fontSize++;
			testFont = new Font("Trebuchet MS", Font.PLAIN, fontSize);
			
			//Get the updated font height
			metrics = factoryFontMetrics(testFont);
			fontHeight = metrics.getHeight();
		}
		
		return font;
	}
	private FontMetrics factoryFontMetrics(Font font)
	{
		//Get the graphics off the main window, because that will always be visible
		Graphics gfx = ScreenCache.getMainScreen().getGraphics();
		return gfx.getFontMetrics(font);
	}

	private void highlightDartboard(Point hoveredPoint)
	{
		DartboardSegmentKt hoveredSegment = getSegmentForPoint(hoveredPoint);
		if (hoveredSegment.equals(lastHoveredSegment))
		{
			//Nothing to do
			return;
		}
		
		if (lastHoveredSegment != null)
		{
			colourSegment(lastHoveredSegment, false);
		}
		
		lastHoveredSegment = hoveredSegment;
		colourSegment(lastHoveredSegment, true);
	}
	private void colourSegment(DartboardSegmentKt segment, boolean highlight)
	{
		if (segment.isMiss())
		{
			//Don't do any highlighting for missing the board
			return;
		}
		
		Color hoveredColour = DartboardUtil.getColourForPointAndSegment(null, segment, highlight, colourWrapper);
		ArrayList<Point> pointsForCurrentSegment = segment.getPoints();
		for (int i=0; i<pointsForCurrentSegment.size(); i++)
		{
			Point pt = pointsForCurrentSegment.get(i);
			colourPoint(pt, hoveredColour);
		}
		
		dartboardLabel.repaint();
	}
	
	private void colourPoint(Point pt, Color colour)
	{
		int x = (int)pt.getX();
		int y = (int)pt.getY();
		
		int rgb = colour.getRGB();
    	int currentRgb = dartboardImage.getRGB(x, y);
    	
    	if (rgb != currentRgb)
    	{
    		dartboardImage.setRGB(x, y, rgb);
    	}
	}
	
	private DartboardSegmentKt getSegmentForPoint(Point pt)
	{
		DartboardSegmentKt segment = hmPointToSegment.get(pt);
		if (segment != null)
		{
			return segment;
		}
		
		Debug.stackTrace("Couldn't find segment for point (" + pt.getX() + ", " + pt.getY() + ")."
					   + "Width = " + getWidth() + ", Height = " + getHeight());
		
		return factoryAndCacheSegmentForPoint(pt);
	}
	
	private DartboardSegmentKt factoryAndCacheSegmentForPoint(Point pt)
	{
		String segmentKey = DartboardUtil.factorySegmentKeyForPoint(pt, centerPt, diameter);
		DartboardSegmentKt segment = hmSegmentKeyToSegment.get(segmentKey);
		if (segment == null)
		{
			segment = new DartboardSegmentKt(segmentKey);
			hmSegmentKeyToSegment.put(segmentKey, segment);
		}
		
		segment.addPoint(pt);
		hmPointToSegment.put(pt, segment);
		
		return segment;
	}
	
	/**
	 * Public methods
	 */
	public ArrayList<Point> getPointsForSegment(int score, int type)
	{
		String segmentKey = score + "_" + type;
		DartboardSegmentKt segment = hmSegmentKeyToSegment.get(segmentKey);
		return segment.getPoints();
	}
	
	public boolean isDouble(Point pt)
	{
		DartboardSegmentKt seg = getSegmentForPoint(pt);
		return seg.isDoubleExcludingBull();
	}
	
	@Override
	public void mouseClicked(MouseEvent arg0) 
	{
		Point pt = arg0.getPoint();
		dartThrown(pt);
	}
	
	public void dartThrown(Point pt)
	{
		Dart dart = convertPointToDart(pt, true);
		if (listener != null)
		{
			addDart(pt);
			
			listener.dartThrown(dart);
		}
	}
	
	public void addOverlay(Point pt, Component overlay)
	{
		add(overlay, 10);
		overlay.setLocation(pt);
	}
	
	public Dart convertPointToDart(Point pt, boolean rationalise)
	{
		if (rationalise)
		{
			rationalisePoint(pt);
		}

		DartboardSegmentKt segment = getSegmentForPoint(pt);
		return DartboardUtil.getDartForSegment(pt, segment);
	}
	
	public void rationalisePoint(Point pt)
	{
		double x = pt.getX();
		double y = pt.getY();
		
		if (x < 0)
		{
			x = 0;
		}
		
		if (y < 0)
		{
			y = 0;
		}
		
		if (x >= getWidth())
		{
			x = getWidth() - 1;
		}
		
		if (y >= getHeight())
		{
			y = getHeight() - 1;
		}
		
		pt.setLocation(x, y);
	}
	
	public double getDiameter()
	{
		return diameter;
	}
	public boolean getRenderScoreLabels()
	{
		return renderScoreLabels;
	}
	public void setRenderScoreLabels(boolean renderScoreLabels)
	{
		this.renderScoreLabels = renderScoreLabels;
	}
	
	public double getArea()
	{
		double radius = diameter / 2;
		return (radius * radius * Math.PI);
	}
	
	public void listen(boolean listen)
	{
		if (listen)
		{
			ensureListening();
		}
		else
		{
			stopListening();
		}
	}
	
	public void ensureListening()
	{
		if (dartboardLabel.getMouseListeners().length == 0)
		{
			dartboardLabel.addMouseListener(this);
			dartboardLabel.addMouseMotionListener(this);
		}
	}
	
	public void stopListening()
	{
		if (dartboardLabel.getMouseListeners().length > 0)
		{
			dartboardLabel.removeMouseListener(this);
			dartboardLabel.removeMouseMotionListener(this);
		}
		
		//Undo any colouring that there might have been
		if (lastHoveredSegment != null)
		{
			colourSegment(lastHoveredSegment, false);
		}
	}
	
	public void doForsyth()
	{
		Random rand = new Random();
		int brucey = rand.nextInt(4) + 1;
		
		doDodgy(ResourceCache.IMG_BRUCE, 300, 478, "forsyth" + brucey);
	}
	
	public void doBull()
	{
		doDodgy(ResourceCache.IMG_DEV, 400, 476, "bull");
	}
	
	public void doBadMiss()
	{
		Random rand = new Random();
		int miss = rand.nextInt(5) + 1;
		
		//4-1 ratio because mitchell > spencer!
		if (miss <= 4)
		{
			doDodgy(ResourceCache.IMG_MITCHELL, 300, 250, "badmiss" + miss);
		}
		else
		{
			doDodgy(ResourceCache.IMG_SPENCER, 460, 490, "damage");
		}
	}
	
	public void doGolfMiss()
	{
		doDodgy(ResourceCache.IMG_DEV, 400, 476, "fourTrimmed");
	}
	
	private void doDodgy(ImageIcon ii, int width, int height, String soundName)
	{
		if (!PreferenceUtil.getBooleanValue(DartsRegistry.PREFERENCES_BOOLEAN_SHOW_ANIMATIONS)
		  || simulation)
		{
			return;
		}
		
		dodgyLabel.setIcon(ii);
		dodgyLabel.setSize(width, height);
		
		int x = (getWidth() - width)/2;
		int y = (getHeight() - height);
		dodgyLabel.setLocation(x, y);
		
		remove(dodgyLabel);
		add(dodgyLabel, 0);
		
		repaint();
		revalidate();
		
		playDodgySound(soundName);
	}
	
	public void playDodgySound(String soundName)
	{
		try
		{
			if (ResourceCache.isInitialised())
			{
				playDodgySoundCached(soundName);
			}
			else
			{
				playDodgySoundAdHoc(soundName);
			}
		}
		catch (Exception e)
		{
			Debug.stackTrace(e, "Caught error playing sound [" + soundName + "]");
		}
	}
	
	@SuppressWarnings("resource")
	private void playDodgySoundCached(String soundName) throws Exception
	{
		AudioInputStream stream = ResourceCache.borrowInputStream(soundName);
		if (stream == null)
		{
			return;
		}
		
		Clip clip = initialiseAudioClip(stream, soundName);
		
		clip.open(stream);
		clip.start();
	}
	
	/**
	 * Old, ad-hoc version for playing sounds (was really slow on home PC).
	 * 
	 * Caches the URL on-the-fly, but still initialises a fresh InputStream every time.
	 */
	@SuppressWarnings("resource")
	private void playDodgySoundAdHoc(String soundName) throws Exception
	{
		URL url = hmSoundNameToUrl.get(soundName);
		if (url == null)
		{
			url = getClass().getResource("/wav/" + soundName + ".wav");
			hmSoundNameToUrl.put(soundName, url);
		}
		
		//Resource may still be null if it genuinely doesn't exist. Just return.
		if (url == null)
		{
			return;
		}
		
		Clip clip = initialiseAudioClip(null, null);

        //This is still slow. Can we cache a 'pool' of the InputStreams? Then rather than close them, call reset() and return them
        //to the pool. Need more than one in case we try to play the same sound twice.
        clip.open(AudioSystem.getAudioInputStream(url));
        clip.start();
	}
	
	private Clip initialiseAudioClip(AudioInputStream stream, String soundName) throws Exception
	{
		//Overwrite the 'clip' variable so this always stores the latest sound. 
		//Allows us to not dismiss the label until the final sound has finished, in the case of overlapping sounds.
		clip = (Clip)AudioSystem.getLine(new Line.Info(Clip.class));
        clip.addLineListener(new LineListener()
        {
            @Override
            public void update(LineEvent event)
            {
            	if (event.getType() == LineEvent.Type.STOP)
                {
            		//Always close or return our one
            		try (Clip myClip = (Clip)event.getLine();)
            		{
            			myClip.stop();
            			
            			if (ResourceCache.isInitialised())
            			{
            				ResourceCache.returnInputStream(soundName, stream);
            			}
            			else
            			{
            				//Close the ad-hoc stream
            				myClip.close();
            			}
            		}
            		
            		//See whether there's currently any clip running. If there isn't, also dismiss our dodgyLabel
            		if (!clip.isRunning())
            		{
            			remove(dodgyLabel);
            	        repaint();
            			revalidate();
            		}
                }
            }
        });
        
        return clip;
	}
	
	private void addDart(Point pt)
	{
		if (dartLabels.isEmpty())
		{
			for (int i=0; i<5; i++)
			{
				JLabel lbl = new JLabel(DARTIMG);
				lbl.setSize(76, 80);
				lbl.setVisible(false);
				add(lbl, 5-i);
				
				dartLabels.add(lbl);
			}
		}
		
		JLabel lbl = dartLabels.get(dartCount);
		lbl.setLocation(pt);
		lbl.setVisible(true);
		dartCount++;
		
		revalidate();
		repaint();
	}
	
	public void clearDarts()
	{
		//Always want to stop this at the same time
		//stopDodgy();
		
		for (int i=0; i<dartLabels.size(); i++)
		{
			JLabel dartLabel = dartLabels.get(i);
			dartLabel.setVisible(false);
		}
		
		dartCount = 0;
		revalidate();
		repaint();
		
	}
	
	public Point getCenterPoint()
	{
		return centerPt;
	}
	
	public void setSimulation(boolean simulation)
	{
		this.simulation = simulation;
	}
	
	@Override
	public void mouseMoved(MouseEvent arg0) 
	{
		highlightDartboard(arg0.getPoint());
	}
	
	@Override
	public void mousePressed(MouseEvent arg0) {}
	@Override
	public void mouseReleased(MouseEvent arg0) {}
	@Override
	public void mouseDragged(MouseEvent arg0) {}
	@Override
	public void mouseEntered(MouseEvent arg0) {}
	@Override
	public void mouseExited(MouseEvent arg0) {}
	
	public BufferedImage factoryOverlay()
	{
		int width = getWidth();
		int height = getHeight();
		Debug.append("Overlay Dimensions: [" + width + ", " + height + "]");
		return new BufferedImage(getWidth(), getHeight(), BufferedImage.TYPE_INT_ARGB);
	}
}

File: src/main/kotlin/dartzee/screen/Dartboard.kt
package dartzee.screen

import dartzee.`object`.ColourWrapper
import dartzee.`object`.Dart
import dartzee.`object`.DartboardSegment
import dartzee.`object`.SegmentType
import dartzee.core.bean.getPointList
import dartzee.core.bean.paint
import dartzee.core.util.getParentWindow
import dartzee.core.util.runOnEventThreadBlocking
import dartzee.listener.DartboardListener
import dartzee.logging.CODE_RENDERED_DARTBOARD
import dartzee.logging.CODE_RENDER_ERROR
import dartzee.logging.KEY_CACHED
import dartzee.logging.KEY_DURATION
import dartzee.screen.game.DartsGameScreen
import dartzee.utils.*
import dartzee.utils.DartsColour.DARTBOARD_BLACK
import dartzee.utils.InjectedThings.logger
import dartzee.utils.ResourceCache.BASE_FONT
import java.awt.*
import java.awt.event.MouseEvent
import java.awt.event.MouseListener
import java.awt.event.MouseMotionListener
import java.awt.image.BufferedImage
import javax.sound.sampled.Clip
import javax.swing.ImageIcon
import javax.swing.JLabel
import javax.swing.JLayeredPane
import javax.swing.SwingConstants

const val LAYER_DARTS = 2
const val LAYER_DODGY = 3
const val LAYER_SLIDER = 4

open class Dartboard(width: Int = 400, height: Int = 400): JLayeredPane(), MouseListener, MouseMotionListener
{
    protected val hmSegmentKeyToSegment = mutableMapOf<String, DartboardSegment>()

    private val dartLabels = mutableListOf<JLabel>()

    private val listeners: MutableList<DartboardListener> = mutableListOf()
    var renderDarts = false
    var centerPoint = Point(200, 200)
        private set

    private var diameter = 360.0

    var scoreLabelColor: Color = Color.WHITE
    var renderScoreLabels = false

    private var dartCount = 0
    var simulation = false

    //Cached things
    private var lastHoveredSegment: DartboardSegment? = null
    private var colourWrapper: ColourWrapper? = null

    //For dodgy sounds/animations
    var latestClip: Clip? = null
    val dodgyLabel = JLabel()

    var dartboardImage: BufferedImage? = null
    val dartboardLabel = JLabel()
     //You know what this is...

    init
    {
        setSize(width, height)
        preferredSize = Dimension(width, height)
        dartboardLabel.setSize(width, height)
        layout = null
        add(dartboardLabel, Integer.valueOf(-1))
    }

    fun addDartboardListener(listener: DartboardListener)
    {
        listeners.add(listener)
    }

    fun paintDartboardCached()
    {
        paintDartboard(cached = true)
    }

    open fun paintDartboard(colourWrapper: ColourWrapper? = null, listen: Boolean = true, cached: Boolean = false)
    {
        val width = width
        val height = height

        val timer = DurationTimer()
        val usingCache = cached && dartboardTemplate != null

        dartboardLabel.setSize(width, height)

        //Initialise/clear down variables
        this.colourWrapper = colourWrapper
        centerPoint = Point(width / 2, height / 2)
        diameter = 0.7 * width

        //Construct the segments, populated with their points. Cache pt -> segment.
        getPointList(width, height).forEach { factoryAndCacheSegmentForPoint(it) }

        if (usingCache)
        {
            initialiseFromTemplate()
        }
        else
        {
            paintDartboardImage()
        }

        addScoreLabels()

        dartboardLabel.icon = ImageIcon(dartboardImage!!)
        dartboardLabel.repaint()

        if (cached
          && dartboardTemplate == null)
        {
            dartboardTemplate = DartboardTemplate(this)
        }

        val duration = timer.getDuration()
        logger.info(CODE_RENDERED_DARTBOARD, "Rendered dartboard[$width, $height] in ${duration}ms (cached: $usingCache)",
            KEY_DURATION to duration, KEY_CACHED to usingCache)

        //Now the dartboard is painted, add the mouse listeners
        if (listen)
        {
            ensureListening()
        }
    }

    private fun paintDartboardImage()
    {
        val hmPointToColor = getAllSegments().flatMap { it.getColorMap(colourWrapper) }.toMap()

        dartboardImage = BufferedImage(width, height, BufferedImage.TYPE_INT_ARGB)
        dartboardImage?.paint { hmPointToColor[it] }
    }

    fun initialiseFromTemplate()
    {
        dartboardImage = dartboardTemplate!!.getDartboardImg()
    }

    private fun addScoreLabels()
    {
        if (!renderScoreLabels)
        {
            return
        }

        //Get the height we want for our labels, which is half the thickness of the outer band
        val radius = diameter / 2
        val outerRadius = UPPER_BOUND_OUTSIDE_BOARD_RATIO * radius
        val lblHeight = Math.round((outerRadius - radius) / 2).toInt()

        val fontToUse = getFontForDartboardLabels(lblHeight)

        for (i in 1..20)
        {
            //Create a label with standard properties
            val lbl = JLabel("" + i)
            lbl.foreground = scoreLabelColor
            lbl.horizontalAlignment = SwingConstants.CENTER
            lbl.font = fontToUse

            //Work out the width for this label, based on the text
            val metrics = factoryFontMetrics(fontToUse)
            val lblWidth = metrics.stringWidth("" + i) + 2
            lbl.setSize(lblWidth, lblHeight)

            //Work out where to place the label
            val points = getPointsForSegment(i, SegmentType.MISS)
            val avgPoint = getAverage(points)
            val lblX = avgPoint.getX().toInt() - lblWidth / 2
            val lblY = avgPoint.getY().toInt() - lblHeight / 2

            val g = dartboardImage!!.graphics as Graphics2D
            g.setRenderingHint(RenderingHints.KEY_ANTIALIASING, RenderingHints.VALUE_ANTIALIAS_ON)
            g.translate(lblX, lblY)
            lbl.paint(g)
        }
    }

    private fun getFontForDartboardLabels(lblHeight: Int): Font
    {
        //Start with a fontSize of 1
        var fontSize = 1f
        var font = BASE_FONT.deriveFont(Font.PLAIN, fontSize)

        //We're going to increment our test font 1 at a time, and keep checking its height
        var testFont = font
        var metrics = factoryFontMetrics(testFont)
        var fontHeight = metrics.height

        while (fontHeight < lblHeight - 2)
        {
            //The last iteration succeeded, so set our return value to be the font we tested.
            font = testFont

            //Create a new testFont, with incremented font size
            fontSize++
            testFont = BASE_FONT.deriveFont(Font.PLAIN, fontSize)

            //Get the updated font height
            metrics = factoryFontMetrics(testFont)
            fontHeight = metrics.height
        }

        return font
    }

    private fun factoryFontMetrics(font: Font): FontMetrics
    {
        //Use a new Canvas rather than going via graphics, as then this will work headless (e.g. from tests)
        return Canvas().getFontMetrics(font)
    }

    fun highlightDartboard(hoveredPoint: Point)
    {
        val hoveredSegment = getSegmentForPoint(hoveredPoint)
        if (hoveredSegment == lastHoveredSegment)
        {
            //Nothing to do
            return
        }

        lastHoveredSegment?.let { colourSegment(it, false) }

        lastHoveredSegment = hoveredSegment
        colourSegment(hoveredSegment, true)
    }

    open fun getInitialColourForSegment(segment: DartboardSegment) = getColourForPointAndSegment(null, segment, colourWrapper)

    fun colourSegment(segment: DartboardSegment, highlight: Boolean)
    {
        if (segment.isMiss())
        {
            return
        }

        val colour = getInitialColourForSegment(segment)
        val hoveredColour = if (highlight && shouldActuallyHighlight(segment)) getHighlightedColour(colour) else colour

        colourSegment(segment, hoveredColour)
    }
    private fun getHighlightedColour(colour: Color): Color =
        if (colour == DARTBOARD_BLACK)
        {
            Color.DARK_GRAY
        }
        else
        {
            DartsColour.getDarkenedColour(colour)
        }

    open fun shouldActuallyHighlight(segment: DartboardSegment) = true

    fun colourSegment(segment: DartboardSegment, col: Color)
    {
        val pointsForCurrentSegment = segment.points
        val edgeColour = getEdgeColourForSegment(segment)
        for (i in pointsForCurrentSegment.indices)
        {
            val pt = pointsForCurrentSegment[i]
            if (edgeColour != null && segment.isEdgePoint(pt))
            {
                colourPoint(pt, edgeColour)
            }
            else
            {
                colourPoint(pt, col)
            }
        }

        dartboardLabel.repaint()
    }

    open fun getEdgeColourForSegment(segment: DartboardSegment) = colourWrapper?.edgeColour

    private fun colourPoint(pt: Point, colour: Color)
    {
        val x = pt.getX().toInt()
        val y = pt.getY().toInt()

        val rgb = colour.rgb
        val currentRgb = dartboardImage!!.getRGB(x, y)

        if (rgb != currentRgb)
        {
            dartboardImage!!.setRGB(x, y, rgb)
        }
    }

    fun getAllSegments() = hmSegmentKeyToSegment.values.toList()

    fun getSegmentForPoint(pt: Point, stackTrace: Boolean = true): DartboardSegment
    {
        val segment = getAllSegments().firstOrNull { it.containsPoint(pt) }
        if (segment != null)
        {
            return segment
        }

        if (stackTrace)
        {
            logger.error(CODE_RENDER_ERROR, "Couldn't find segment for point (" + pt.getX() + ", " + pt.getY() + ")."
                    + "Width = " + width + ", Height = " + height)
        }

        return factoryAndCacheSegmentForPoint(pt)
    }

    private fun factoryAndCacheSegmentForPoint(pt: Point): DartboardSegment
    {
        val newSegment = factorySegmentForPoint(pt, centerPoint, diameter)
        val segmentKey = "${newSegment.score}_${newSegment.type}"

        val segment = hmSegmentKeyToSegment.getOrPut(segmentKey) { newSegment }
        segment.addPoint(pt)
        return segment
    }

    /**
     * Public methods
     */
    fun getPointsForSegment(score: Int, type: SegmentType): MutableList<Point>
    {
        val segmentKey = score.toString() + "_" + type
        val segment = hmSegmentKeyToSegment[segmentKey]
        return segment?.points ?: mutableListOf()
    }
    fun getSegment(score: Int, type: SegmentType): DartboardSegment? = hmSegmentKeyToSegment["${score}_$type"]

    fun isDouble(pt: Point): Boolean
    {
        val seg = getSegmentForPoint(pt)
        return seg.isDoubleExcludingBull()
    }

    fun getPotentialAimPoints() = getPotentialAimPoints(centerPoint, diameter)
    fun translateAimPoint(aimPoint: AimPoint) = AimPoint(centerPoint, diameter / 2, aimPoint.angle, aimPoint.ratio).point

    open fun dartThrown(pt: Point)
    {
        val dart = convertPointToDart(pt, true)

        if (renderDarts)
        {
            runOnEventThreadBlocking { addDart(pt) }
        }

        listeners.forEach {
            it.dartThrown(dart)
        }
    }

    fun addOverlay(pt: Point, overlay: Component)
    {
        add(overlay)
        setLayer(overlay, LAYER_SLIDER)
        overlay.location = pt
    }

    fun convertPointToDart(pt: Point, rationalise: Boolean): Dart
    {
        if (rationalise)
        {
            rationalisePoint(pt)
        }

        val segment = getSegmentForPoint(pt)
        return getDartForSegment(pt, segment)
    }

    fun rationalisePoint(pt: Point)
    {
        val x = pt.x.coerceIn(0, width - 1)
        val y = pt.y.coerceIn(0, height - 1)

        pt.setLocation(x, y)
    }

    fun listen(listen: Boolean)
    {
        if (listen)
        {
            ensureListening()
        }
        else
        {
            stopListening()
        }
    }

    fun ensureListening()
    {
        if (dartboardLabel.mouseListeners.isEmpty())
        {
            dartboardLabel.addMouseListener(this)
            dartboardLabel.addMouseMotionListener(this)
        }
    }

    fun stopListening()
    {
        if (dartboardLabel.mouseListeners.isNotEmpty())
        {
            dartboardLabel.removeMouseListener(this)
            dartboardLabel.removeMouseMotionListener(this)
        }

        //Undo any colouring that there might have been
        lastHoveredSegment?.let { colourSegment(it, false) }
    }

    private fun addDart(pt: Point)
    {
        if (dartLabels.isEmpty())
        {
            for (i in 0..4)
            {
                val lbl = JLabel(DARTIMG)
                lbl.setSize(76, 80)
                lbl.isVisible = false
                add(lbl)

                dartLabels.add(lbl)
            }
        }

        val lbl = dartLabels[dartCount]
        lbl.location = pt
        lbl.isVisible = true
        dartCount++

        setLayer(lbl, LAYER_DARTS, 5-dartCount)

        revalidate()
        repaint()
    }

    fun clearDarts()
    {
        //Always want to stop this at the same time
        //stopDodgy();

        for (i in dartLabels.indices)
        {
            val dartLabel = dartLabels[i]
            dartLabel.isVisible = false
        }

        dartCount = 0
        revalidate()
        repaint()
    }

    override fun mouseMoved(arg0: MouseEvent)
    {
        if (getParentWindow()?.isFocused == true)
        {
            highlightDartboard(arg0.point)
        }
    }

    override fun mouseReleased(arg0: MouseEvent)
    {
        if (!suppressClickForGameWindow())
        {
            val pt = arg0.point
            dartThrown(pt)
        }
    }
    private fun suppressClickForGameWindow(): Boolean
    {
        val scrn = getParentWindow() as? DartsGameScreen ?: return false
        if (scrn.haveLostFocus)
        {
            scrn.haveLostFocus = false
            return true
        }

        return false
    }

    override fun mouseClicked(arg0: MouseEvent) {}
    override fun mousePressed(arg0: MouseEvent) {}
    override fun mouseDragged(arg0: MouseEvent) {}
    override fun mouseEntered(arg0: MouseEvent) {}
    override fun mouseExited(arg0: MouseEvent) {}

    fun factoryOverlay() = BufferedImage(width, height, BufferedImage.TYPE_INT_ARGB)

    companion object
    {
        private val DARTIMG = ImageIcon(Dartboard::class.java.getResource("/dartImage.png"))
        var dartboardTemplate: DartboardTemplate? = null

        fun appearancePreferenceChanged()
        {
            dartboardTemplate = null
        }
    }

    inner class DartboardTemplate(dartboard: Dartboard)
    {
        private val dartboardImg = dartboard.dartboardImage!!

        fun getDartboardImg(): BufferedImage
        {
            val cm = dartboardImg.colorModel
            val isAlphaPremultiplied = cm.isAlphaPremultiplied
            val writableRaster = dartboardImg.raster.createCompatibleWritableRaster()
            val raster = dartboardImg.copyData(writableRaster)
            return BufferedImage(cm, raster, isAlphaPremultiplied, null)
        }

    }

}
